[WIP] Arellano bond panel data

jaimeoliver1 opened the pull request:
Hi, 

we have implemented the Arellano Bond estimator based on this previous implementation: https://gist.github.com/toobaz/6d01be09bbd56de7038fe8ff26780138. 
We have two comments: 
- The AB estimator is a panel data model, but it builds on top of an IV model. Therefore, we have thought that importing `IVGMM` from the `linearmodels.iv.model` sub-module was the easiest option, but we are open to other options. 
- We have not implemented any further tests, given that the model builds on top of already tested models. Any insight on what would the testing strategy be here?


This PR closes https://github.com/bashtage/linearmodels/issues/95

pep8speaks commented:
Hello @jaimeoliver1! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`linearmodels/panel/model.py`](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py):

> [Line 3143:57](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3143): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\e'
> [Line 3155:25](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3155): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 3155:26](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3155): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 3156:25](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3156): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 3156:26](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3156): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 3208:42](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3208): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 3209:52](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3209): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 3223:100](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3223): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (122 > 99 characters)
> [Line 3226:42](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3226): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 3226:89](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3226): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 3226:91](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3226): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 3226:100](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3226): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (182 > 99 characters)
> [Line 3226:111](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3226): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 3226:112](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3226): [E262](https://duckduckgo.com/?q=pep8%20E262) inline comment should start with '# '
> [Line 3231:100](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3231): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 99 characters)
> [Line 3233:100](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3233): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 99 characters)
> [Line 3250:100](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3250): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 99 characters)
> [Line 3252:100](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3252): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 99 characters)
> [Line 3260:100](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3260): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 3274:100](https://github.com/bashtage/linearmodels/blob/06a895a4ee4993c4d4aa464b7c9caa78e14e2383/linearmodels/panel/model.py#L3274): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 99 characters)

bashtage commented:
Can you run black and isort?

jaimeoliver1 commented:
@bashtage I committed the file with the correct formatting

bashtage commented:
I think the best way to test would be to take the Stata code as a reference and test the various outputs and options.

bashtage commented:
Overall looks very good. 

1. Need to add it to the docs. 
2. Can you break these long comments that are making PEP unhappy.
3. Can you use a raw string `r"Code with escape like \epsilon"` to silence the other.

jaimeoliver1 commented:
Also, we fixed the linting error and silence the  "\epsilon" issues. 
We also added documentation in the docstring. However, doc build is failing (even in master...)

bashtage commented:
Thanks for following up.  I can run Stata and generate results.

bashtage commented:
Going to close and reopen to trigger a rebuild.
